### ISSUE TITLE
Support docker swarm over Unix socket

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#7603]
Just tested this with the current RC:

I would suggest that the "host" parameter also supports unix-sockets. Exposing the docker socket over TCP is not something you would usually do in a production environment. Granting other containers access to the docker management port is commonly done by mounting /var/run/docker.sock read-only inside the container.

_Originally posted by @darkl0rd in https://github.com/prometheus/prometheus/pull/7420#issuecomment-660109510_

**Comment by brian-brazil** [comment]
Would that require you to be on a very specific machine, and thus not work with HA setups? I think we should not be doing this, for the same reason this SD doesn't default to talking to localhost.

**Comment by darkl0rd** [comment]
2375 is by default not exposed, nor is 2376. Since 19.03 this is also considered malpractice, as docker can be managed through SSH (DOCKER_HOST=ssh://manager-node.domain) - which means you no longer need to expose the socket insecurely (2375) or setup a separate PKI (2376) but rather securely manage your Swarm using your existing PKI (SSH). Effectively the Docker client establishes an SSH session to the ${DOCKER_HOST} and communicates with the Unix socket (/var/run/docker.sock) of that host.

When running Prometheus on a manager node, be it containerised or as a regular process, it would be beneficial if one could directly access file:///var/run/docker.sock rather than having to expose the (secure) management port.

The resulting prometheus configuration would then be:
```
scrape_configs:
  # Create a job for Docker Swarm containers.
  - job_name: 'swarm'
    dockerswarm_sd_configs:
      - host: file:///var/run/docker.sock
        role: tasks
```

**Comment by darkl0rd** [comment]
> Would that require you to be on a very specific machine, and thus not work with HA setups? I think we should not be doing this, for the same reason this SD doesn't default to talking to localhost.

Yes, this would require you to deploy prometheus on a docker "manager" node - which is in the world of swarm common practice for those processes that need access to the management port (Likewise in k8s).

**Comment by roidelapluie** [comment]
I was assuming that swarm only allowed http. It turns out it is because we pass a http client. See #7604 for a fix.

**Comment by roidelapluie** [comment]
> Would that require you to be on a very specific machine, and thus not work with HA setups? I think we should not be doing this, for the same reason this SD doesn't default to talking to localhost.

I do not think lot of people would put the docker api behind a load balancer anyway. Using the socket does not mean that it would not work on HA setups.

**Comment by brian-brazil** [comment]
That doesn't sound like a SPOF, so that sounds okay.

**Comment by beorn7** [comment]
Fixed in release-2.20 (to be merged into master very soon).
